File: csharp/Utility.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
public class Utility
{
	private static readonly HashSet<char> charsWithQuoteSet = new HashSet<char> { ',', '\n', '\r', '\"' };

	public Utility() { }

	public virtual string GetMasterDataFilePath()
	{
		return "";
	}

	public virtual string ExtractOtp(string dataRow)
	{
		return "";
	}

	public string[] ReadFileData()
	{
		List<List<string>> dataRows = new List<List<string>>();
		string filePath = GetMasterDataFilePath();

		if (!string.IsNullOrEmpty(filePath) && File.Exists(filePath))
		{
			try
			{
				using (StreamReader csvReader = new StreamReader(filePath))
				{
					string header = csvReader.ReadLine(); // Skip header
					List<string> columnValues = new List<string>();

					while (TryGetDataRow(csvReader, true, columnValues))
					{
						columnValues.RemoveAt(columnValues.Count - 1);
						dataRows.Add(new List<string>(columnValues));
						columnValues = new List<string>();
					}
				}
			}
			catch (IOException e)
			{
				Console.WriteLine(e.Message);
			}
		}

		return dataRows.SelectMany(row => row).ToArray();
	}

	public static bool TryGetDataRow(StreamReader csvReader, bool skipEmptyRows, List<string> columnValues)
	{
		columnValues.Clear();

		if (csvReader != null && !csvReader.EndOfStream)
		{
			bool isCurrentRowEmpty;
			do
			{
				columnValues.Clear();
				bool openingQuote = false;
				StringBuilder sb = new StringBuilder();

				int intChar;
				while ((intChar = csvReader.Read()) != -1)
				{
					char currentChar = (char)intChar;

					if (currentChar == '\"')
					{
						openingQuote = !openingQuote;
					}

					if (!openingQuote && currentChar == ',')
					{
						columnValues.Add(sb.ToString());
						sb.Clear();
					}
					else if (!openingQuote && (currentChar == '\n' || currentChar == '\r'))
					{
						int nextIntChar = csvReader.Peek();
						if (nextIntChar != -1)
						{
							char nextChar = (char)nextIntChar;
							if (nextChar == '\n' || nextChar == '\r')
							{
								csvReader.Read(); 
							}
						}

						if (sb.Length > 0 || columnValues.Count > 0)
						{
							columnValues.Add(sb.ToString());
						}
						break;
					}
					else
					{
						sb.Append(currentChar);
					}

					if (csvReader.EndOfStream)
					{
						columnValues.Add(sb.ToString());
					}
				}

				isCurrentRowEmpty = true;
				for (int i = 0; i < columnValues.Count; i++)
				{
					string itemValue = columnValues[i];

					if (!string.IsNullOrEmpty(itemValue))
					{
						int startIndex = itemValue.StartsWith("\"") ? 1 : 0;
						int endIndex = itemValue.EndsWith("\"") ? itemValue.Length - 1 : itemValue.Length;

						StringBuilder processedValue = new StringBuilder();
						int quoteIndex = -1;

						for (int j = startIndex; j < endIndex; j++)
						{
							if (itemValue[j] == '\"')
							{
								if (j - quoteIndex != 1)
								{
									processedValue.Append('\"');
								}
								quoteIndex = j;
							}
							else
							{
								processedValue.Append(itemValue[j]);
							}
						}

						columnValues[i] = processedValue.ToString().Trim();
					}

					if (!string.IsNullOrEmpty(columnValues[i]))
					{
						isCurrentRowEmpty = false;
					}
				}

				if (skipEmptyRows && isCurrentRowEmpty)
				{
					columnValues.Clear();
				}

			} while (skipEmptyRows && isCurrentRowEmpty && !csvReader.EndOfStream);
		}

		return columnValues.Count > 0;
	}

	public void WriteFile(string[][] dataRows)
	{
		List<List<string>> dataRow = new List<List<string>>();
		string filePath = GetMasterDataFilePath();

		if (!string.IsNullOrEmpty(filePath) && File.Exists(filePath))
		{
			try
			{
				using (StreamReader csvReader = new StreamReader(filePath))
				{
					csvReader.ReadLine(); // Skip header
					List<string> columnValues = new List<string>();

					while (TryGetDataRow(csvReader, true, columnValues))
					{
						dataRow.Add(new List<string>(columnValues));
						columnValues = new List<string>();
					}
				}
			}
			catch (IOException e)
			{
				Console.WriteLine(e.Message);
			}
		}

		string[][] res = dataRow.Select(row => row.ToArray()).ToArray();

		FileInfo originalFile = new FileInfo(filePath);
		string folderPath = originalFile.DirectoryName;

		string outputFilePath = string.IsNullOrEmpty(folderPath) ? "output.csv" : Path.Combine(folderPath, "output.csv");
		string failedCasesFilePath = string.IsNullOrEmpty(folderPath) ? "failedCases.csv" : Path.Combine(folderPath, "failedCases.csv");

		int failCases = 0;

		try
		{
			using (StreamWriter csvWriter = new StreamWriter(outputFilePath))
			using (StreamWriter csvWriter_2 = new StreamWriter(failedCasesFilePath))
			{
				csvWriter.WriteLine("Message,Extracted,Expected");
				csvWriter_2.WriteLine("Message,Extracted,Expected");

				for (int it = 0; it < dataRows.Length; it++)
				{
					bool flag = false;
					if (dataRows[it].Length != 2 || !dataRows[it][1].Equals(res[it][1], StringComparison.OrdinalIgnoreCase))
					{
						failCases++;
						flag = true;
					}

					List<string> rows = new List<string>();
					foreach (string data in dataRows[it])
					{
						StringBuilder sb = new StringBuilder();
						bool specialCharFound = false;

						foreach (char currentChar in data)
						{
							sb.Append(currentChar);

							if (currentChar == '\"')
							{
								sb.Append(currentChar);
							}

							if (charsWithQuoteSet.Contains(currentChar))
							{
								specialCharFound = true;
							}
						}

						if (specialCharFound)
						{
							sb.Insert(0, '\"').Append('\"');
						}

						rows.Add(sb.ToString());
					}

					rows.Add(res[it][1]);

					csvWriter.WriteLine(string.Join(",", rows));
					if (flag)
					{
						csvWriter_2.WriteLine(string.Join(",", rows));
					}
				}

				double accuracy = ((double)(dataRows.Length - failCases) / dataRows.Length) * 100;
				Console.WriteLine("Accuracy: " + accuracy + "%");
				Console.WriteLine("CSV writing completed successfully: " + outputFilePath);
			}
		}
		catch (IOException ex)
		{
			Console.WriteLine("Error writing CSV: " + ex.Message);
		}
	}
}

File: java/Utility.java
import java.io.*;
import java.util.*;

public class Utility {

    private static final Set<Character> charsWithQuoteSet = new HashSet<>(Arrays.asList(',', '\n', '\r', '\"'));

    public Utility(){}

    public String[] readFileData() {
        List<List<String>> dataRows = new ArrayList<>();
        String filePath = getMasterDataFilePath();

        if (filePath != null && !filePath.isEmpty() && new File(filePath).exists()) {
            try {
                BufferedReader csvReader = new BufferedReader(new FileReader(filePath));
                List<String> columnValues = new ArrayList<>();
                csvReader.readLine(); // Skip header
                while (tryGetDataRow(csvReader, true, columnValues)) {
                    columnValues.remove(columnValues.size() - 1);
                    dataRows.add(columnValues);
                    columnValues = new ArrayList<>();
                }
            } catch (IOException e) {
                System.out.println(e.getMessage());
            }
        }

        String[] res = dataRows.stream().flatMap(row -> row.stream()).toArray(size -> new String[size]);
        return res;
    }

    public static boolean tryGetDataRow(BufferedReader csvReader, boolean skipEmptyRows, java.util.List<String> columnValues) throws java.io.IOException {
        columnValues.clear();

        if (csvReader != null && csvReader.ready()) {
            boolean isCurrentRowEmpty;
            do {
                columnValues.clear();
                boolean openingQuote = false;
                StringBuilder sb = new StringBuilder();

                int intChar;
                while ((intChar = csvReader.read()) != -1) {
                    char currentChar = (char) intChar;

                    if (currentChar == '\"') {
                        openingQuote = !openingQuote;
                    }

                    if (!openingQuote && currentChar == ',') {
                        columnValues.add(sb.toString());
                        sb.setLength(0);
                    } else if (!openingQuote && (currentChar == '\n' || currentChar == '\r')) {
                        int nextIntChar = csvReader.read();
                        if (nextIntChar != -1) {
                            char nextChar = (char) nextIntChar;
                            if (nextChar != currentChar && (nextChar == '\n' || nextChar == '\r')) {
                                // skip second newline char in CRLF or LFCR
                            } else {
                                csvReader.reset();
                            }
                        }
                        if (sb.length() > 0 || !columnValues.isEmpty()) {
                            columnValues.add(sb.toString());
                        }
                        break;
                    } else {
                        sb.append(currentChar);
                    }

                    if (csvReader.markSupported()) {
                        csvReader.mark(1);
                    }
                    if (!csvReader.ready()) {
                        columnValues.add(sb.toString());
                    }
                }

                isCurrentRowEmpty = true;
                for (int i = 0; i < columnValues.size(); i++) {
                    String itemValue = columnValues.get(i);

                    if (!itemValue.isEmpty()) {
                        int startIndex = itemValue.charAt(0) == '\"' ? 1 : 0;
                        int endIndex = itemValue.length() - (itemValue.charAt(itemValue.length() - 1) == '\"' ? 1 : 0);

                        StringBuilder processedValue = new StringBuilder();
                        int quoteIndex = -1;
                        for (int j = startIndex; j < endIndex; j++) {
                            if (itemValue.charAt(j) == '\"') {
                                if (j - quoteIndex != 1) {
                                    processedValue.append('\"');
                                }
                                quoteIndex = j;
                            } else {
                                processedValue.append(itemValue.charAt(j));
                            }
                        }

                        columnValues.set(i, processedValue.toString().trim());
                    }

                    if (!columnValues.get(i).isEmpty()) {
                        isCurrentRowEmpty = false;
                    }
                }

                if (skipEmptyRows && isCurrentRowEmpty) {
                    columnValues.clear();
                }
            } while (skipEmptyRows && isCurrentRowEmpty && csvReader.ready());
        }

        return !columnValues.isEmpty();
    }

    public void writeFile(String[][] dataRows) {
        List<List<String>> dataRow = new ArrayList<>();
        String filePath = getMasterDataFilePath();

        if (filePath != null && !filePath.isEmpty() && new File(filePath).exists()) {
            try {
                BufferedReader csvReader = new BufferedReader(new FileReader(filePath));
                List<String> columnValues = new ArrayList<>();
                csvReader.readLine(); // Skip header
                while (tryGetDataRow(csvReader, true, columnValues)) {
                    dataRow.add(columnValues);
                    columnValues = new ArrayList<>();
                }
            } catch (IOException e) {
                System.out.println(e.getMessage());
            }
        }

        String[][] res = dataRow.stream().map(row -> row.toArray(new String[0])).toArray(size -> new String[size][]);

        File originalFile = new File(filePath);
        String folderPath = originalFile.getParent();
        String outputFilePath ="";
        String failedCasesFilePath ="";
        if(folderPath==null){
            outputFilePath ="output.csv";
            failedCasesFilePath ="failedCases.csv";
        }
        else{
            outputFilePath =folderPath + File.separator + "output.csv";
            failedCasesFilePath =folderPath + File.separator + "failedCases.csv";
        }

        int failCases = 0;
        BufferedWriter csvWriter = null;
        BufferedWriter csvWriter_2= null;

        try {
            csvWriter = new BufferedWriter(new FileWriter(outputFilePath));
            csvWriter.write("Message,Extracted,Expected\n");

            csvWriter_2 = new BufferedWriter(new FileWriter(failedCasesFilePath));
            csvWriter_2.write("Message,Extracted,Expected\n");

            for (int it = 0; it < dataRows.length; it++) {
                boolean flag=false;
                if (dataRows[it].length != 2 || !dataRows[it][1].equalsIgnoreCase(res[it][1])) {
                    failCases++;
                    flag=true;
                }

                List<String> rows = new ArrayList<>();
                for (String data : dataRows[it]) {
                    StringBuilder sb = new StringBuilder();
                    boolean specialCharFound = false;

                    for (int i = 0; i < data.length(); i++) {
                        char currentChar = data.charAt(i);
                        sb.append(currentChar);

                        if (currentChar == '\"') {
                            sb.append(currentChar);
                        }

                        if (charsWithQuoteSet.contains(currentChar)) {
                            specialCharFound = true;
                        }
                    }

                    if (specialCharFound) {
                        sb.insert(0, '\"').append('\"');
                    }

                    rows.add(sb.toString());

                }

                rows.add(res[it][1]);

                csvWriter.write(String.join(",", rows));
                csvWriter.newLine();

                if(flag){
                    csvWriter_2.write(String.join(",", rows));
                    csvWriter_2.newLine();
                }
            }

            double accuracy = (dataRows.length - failCases) / (double) dataRows.length * 100;
            System.out.println("Accuracy: " + accuracy + "%");
            System.out.println("CSV writing completed successfully: " + outputFilePath);

        } catch (IOException ex) {
            System.out.println("Error writing CSV: " + ex.getMessage());
        } finally {
            try {
                if (csvWriter != null) {
                    csvWriter.close();
                }
                if (csvWriter_2 != null) {
                    csvWriter_2.close();
                }
            } catch (IOException e) {
                System.out.println("Error closing CSV writer: " + e.getMessage());
            }
        }
    }

    public String extractOtp(String dataRow) {
        return "";
    }

    public String getMasterDataFilePath() {
        return "";
    }
}
